### PR TITLE
StatusGrid refreshes to show the latest information

### LINKED ISSUE #422: Status Grid
Flutter widget that takes in results for commits, and shows them in a matrix. Should be able to scroll horizontally and vertically to show overflow.

## Milestones
- [X] Widget that can show a grid
- [X] Widget takes in List<CommitStatus> and shows it in the grid (CommitStatus imported from `app_dart/`)
- [ ] Widget is stateful and passes updates down to child widgets

### PR DESCRIPTION
**Pull request by CaseyHillers**
Implemented the Provider package to handle the state management for the new build dashboard. Created a state class for the build dashboard. Once feature parity is reached with the current dashboard, it will contain more variables (e.g. tree build status, agent statuses).

Created a new wrapper widget for the StatusGrid that handles the state management section for now. If we find this creates performance issues, we can refactor it.

This fixes #422 since the feature parity is met. The dashboard loads production data in, and updates it with the latest information on a set interval.

## Tested
I did a manual test by building the app for release on my phone, and seeing if tasks that were running went to a different state.

### CODE REVIEW

File: app_flutter/lib/state/flutter_build.dart [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+
+import 'package:cocoon_service/protos.dart' show CommitStatus;
+
+import '../service/cocoon.dart';
+
+/// State for the Flutter Build Dashboard
+class FlutterBuildState extends ChangeNotifier {
```
Comment (jonahwilliams) [inline app_flutter/lib/state/flutter_build.dart]: This class should have a `dispose` method that cancels the periodic timer. Then in the _BuildStatusPageState, you can call that method in the widget's `dispose`

File: app_flutter/lib/status_grid.dart [hunk 2]
```diff
@@ -3,43 +3,61 @@
...
 
+import 'state/flutter_build.dart';
 import 'commit_box.dart';
 import 'task_box.dart';
 
+/// Container that manages the layout and data handling for [StatusGrid].
+///
+/// If there's no data for [StatusGrid], it shows [CircularProgressIndicator].
+class StatusGridContainer extends StatelessWidget {
+  @override
```
Comment (jonahwilliams) [inline app_flutter/lib/status_grid.dart]: This class can have a `const` constructor

File: app_flutter/lib/status_grid.dart [hunk 3]
```diff
@@ -3,43 +3,61 @@
...
     show Commit, CommitStatus, Stage, Task;
 
+import 'state/flutter_build.dart';
 import 'commit_box.dart';
 import 'task_box.dart';
 
+/// Container that manages the layout and data handling for [StatusGrid].
+///
+/// If there's no data for [StatusGrid], it shows [CircularProgressIndicator].
+class StatusGridContainer extends StatelessWidget {
```
Comment (jonahwilliams) [inline app_flutter/lib/status_grid.dart]: Are there existing tests that cover this class? Would like to have a test that asserts that the progress indicator display when status is empty is correct

File: app_flutter/lib/state/flutter_build.dart (hunk 1 above)
Comment (CaseyHillers) [reply app_flutter/lib/state/flutter_build.dart]: Added

File: app_flutter/lib/status_grid.dart (hunk 3 above)
Comment (CaseyHillers) [reply app_flutter/lib/status_grid.dart]: Yes, they're in `status_grid_test.dart` and cover that case. Should they be moved to `status_grid_container_test.dart`?

File: app_flutter/lib/state/flutter_build.dart [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  /// Timer that calls [_fetchBuildStatusUpdate] on a set interval.
+  Timer _refreshTimer;
+
+  /// The current status of the commits loaded.
+  List<CommitStatus> statuses = [];
+
+  /// Start a fixed interval loop that fetches build state updates based on [_refreshRate].
+  void startFetchingBuildStateUpdates() async {
+    if (_refreshTimer != null) {
+      throw 'already fetching build state updates';
```
Comment (jonahwilliams) [inline app_flutter/lib/state/flutter_build.dart]: While it's not a compile error to throw a String, generally you should only throw `Exception` or `Error`. That said, I don't think this should throw at all, just return if it is already set.

File: app_flutter/lib/status_grid.dart (hunk 2 above)
Comment (CaseyHillers) [reply app_flutter/lib/status_grid.dart]: Added!

File: app_flutter/lib/state/flutter_build.dart (hunk 4 above)
Comment (CaseyHillers) [reply app_flutter/lib/state/flutter_build.dart]: Done. I modified the test to check if the `refreshTimer` is the same across multiple start calls.

Thanks for sharing the use about `Exception` and `Error`. I saw the backend doing it that way so I assumed that was the better practice :)

File: app_flutter/lib/status_grid.dart (hunk 3 above)
Comment (jonahwilliams) [reply app_flutter/lib/status_grid.dart]: Nah, that is fine

File: app_flutter/lib/state/flutter_build.dart [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      return;
+    }
+
+    refreshTimer =
+        Timer.periodic(_refreshRate, (t) => _fetchBuildStatusUpdate());
+  }
+
+  /// Request the latest [statuses] from [CocoonService].
+  void _fetchBuildStatusUpdate() async {
+    statuses = await _cocoonService.fetchCommitStatuses();
```
Comment (jonahwilliams) [inline app_flutter/lib/state/flutter_build.dart]: Would be nice to test that this gets called without waiting 10 seconds

File: app_flutter/lib/state/flutter_build.dart [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    notifyListeners();
+  }
```
Comment (digiter) [inline app_flutter/lib/state/flutter_build.dart]: What happens when `fetchCommitStatuses` returns a HttpException?

File: app_flutter/lib/state/flutter_build.dart (hunk 5 above)
Comment (CaseyHillers) [reply app_flutter/lib/state/flutter_build.dart]: Would `WidgetTester.pump(Duration)` work for this?

File: app_flutter/lib/state/flutter_build.dart (hunk 6 above)
Comment (CaseyHillers) [reply app_flutter/lib/state/flutter_build.dart]: That's a good point. Do you have any recommendations on how to forward this error to the widget?

My current approach would be `Stream`, but that seems to over engineer what we want to do here. A naive approach would be to just assume this was a one off incident, and don't update statuses.

File: app_flutter/lib/state/flutter_build.dart (hunk 5 above)
Comment (jonahwilliams) [reply app_flutter/lib/state/flutter_build.dart]: Maybe...

File: app_flutter/test/service/mock_cocoon.dart [hunk 7]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:app_flutter/service/fake_cocoon.dart';
+import 'package:mockito/mockito.dart';
+
+/// CocoonService for checking interactions.
+class MockCocoonService extends Mock implements FakeCocoonService {}
```
Comment (jonahwilliams) [inline app_flutter/test/service/mock_cocoon.dart]: Its fairly common to leave mocks as a single line declaration at the bottom of the test they are used in

File: app_flutter/lib/state/flutter_build.dart (hunk 5 above)
Comment (CaseyHillers) [reply app_flutter/lib/state/flutter_build.dart]: https://github.com/flutter/cocoon/blob/411ae1a428db2c445da6a2dfc7ee5de5ceb5d195/app_flutter/test/state/flutter_build_test.dart#L24

I tried a basic test here, but it throws an exception since the refreshTimer is still running at the end of the test. On tear down, I call FlutterBuildState.dispose() that cancels the timer.

File: app_flutter/test/service/mock_cocoon.dart (hunk 7 above)
Comment (CaseyHillers) [reply app_flutter/test/service/mock_cocoon.dart]: If this gets used in another file (maybe in another PR), would I move it to its own file?

Comment (jonahwilliams) [reply app_flutter/test/service/mock_cocoon.dart]: Not generally, since it so small

File: app_flutter/lib/state/flutter_build.dart (hunk 5 above)
Comment (jonahwilliams) [reply app_flutter/lib/state/flutter_build.dart]: I would tear down the buildState in the test and not tearDown so the test can pass then

File: app_flutter/lib/state/flutter_build.dart (hunk 6 above)
Comment (CaseyHillers) [reply app_flutter/lib/state/flutter_build.dart]: We talked in person and I created https://github.com/flutter/cocoon/issues/451 that sums up our solution.

File: app_flutter/lib/state/flutter_build.dart (hunk 5 above)
Comment (CaseyHillers) [reply app_flutter/lib/state/flutter_build.dart]: Okay that worked. That's an interesting issue.

The test I implemented takes less than a few seconds to run, so I think this is done.

Comment (jonahwilliams) [reply app_flutter/lib/state/flutter_build.dart]: The widget tests should be running in a fake async zone that controls the flow of time
